### PR TITLE
Refactor SlsVersion

### PR DESCRIPTION
**Pull request by ferozco**
In this PR:
- Refactor SlsVersion to provide consistent API
- Remove verbose tests
- Fix issues with SlsVersionMatcher that were discovered during the refactor

Ex:
```typescript
const slsVersion = SlsVersion.safeValueOf("1.2.0");
const slsVersionMatcher = SlsVersionMatcher.safeValueOf("1.2.x");

slsVersionMatcher.matches(slsVersion); // true
slsVersion.compare(SlsVersion.of("2.0.0"); // -1

```

### CODE REVIEW

File: src/slsVersion.ts [hunk 1]
```diff
@@ -32,127 +32,123 @@ export interface ISlsVersion {
...
-/** Parse an SLS version string. */
-export function parse(version: string): ISlsVersion {
-    // Parse the version string as an SLS version
-    const match = version.match(/^([0-9]+)\.([0-9]+)\.([0-9]+)(?:-rc([0-9]+))?(?:-([0-9]+)-g([a-f0-9]+))?(\.dirty)?$/);
+export class SlsVersion implements ISlsVersion {
+    private static MATCHER = /^([0-9]+)\.([0-9]+)\.([0-9]+)(?:-rc([0-9]+))?(?:-([0-9]+)-g([a-f0-9]+))?(\.dirty)?$/;
 
-    if (null == match) {
-        throw new Error(`Invalid SLS version: "${version}"`);
+    public static safeValueOf(version: string): SlsVersion | null {
```
Comment (ferozco) [inline src/slsVersion.ts]: maybe used undefined

File: src/slsVersion.ts [hunk 2]
```diff
@@ -32,127 +32,123 @@ export interface ISlsVersion {
...
-    } else if (isReleaseCandidate(lhs) && isReleaseCandidate(rhs) && lhs.rc !== rhs.rc) {
-        return lhs.rc > rhs.rc;
-    } else if (isReleaseCandidateSnapshot(lhs) && isNormalReleaseCandidate(rhs)) {
-        return true;
-    } else if (isReleaseCandidateSnapshot(lhs) && isReleaseCandidateSnapshot(rhs)) {
-        return lhs.snapshot > rhs.snapshot;
-    } else {
-        return false;
+    public static isValid(version: string): boolean {
+        try {
```
Comment (ferozco) [inline src/slsVersion.ts]: avoid using exceptions

File: src/slsVersion.ts [hunk 3]
```diff
@@ -32,127 +32,123 @@ export interface ISlsVersion {
...
+        private readonly value: string,
+        public readonly major: number,
+        public readonly minor: number,
+        public readonly patch: number,
+        public readonly rc: number | undefined,
+        public readonly snapshot: number | undefined,
+        public readonly orderable: boolean,
+    ) {}
+
+    public compare(other: ISlsVersion): -1 | 0 | 1 {
```
Comment (ferozco) [inline src/slsVersion.ts:100]: new enum

File: src/slsVersionMatcher.ts [hunk 4]
```diff
@@ -96,26 +102,10 @@ export class SlsVersionMatcher implements ISlsVersionMatcher {
         return this.compare(version) === 0;
     }
 
-    public compare(version: ISlsVersion) {
+    public compare(version: ISlsVersion): -1 | 0 | 1 {
```
Comment (ferozco) [inline src/slsVersionMatcher.ts:107]: comparitor?

File: src/slsVersion.ts (hunk 3 above)
Comment (styu) [reply src/slsVersion.ts:100]: why an enum? i think -1/0/1 are pretty standard for comparison results

**Review by styu — APPROVED** [review]

File: src/slsVersion.ts (hunk 3 above)
Comment (iamdanfox) [reply src/slsVersion.ts:100]: That was a request from me- I thought a TS enum where the values are actually the numbers would make the code extra readable: 

enum Comparison {
  GREATER = 1
  EQUAL = 0
  LESS_THAN =-1
}

I know 1,-1 etc is common, just wondering if we can be better

**Pull request merged**